### DISCUSSION TITLE
Fournir une API, ou plutôt un kit d'hébergement d'API ?

### DISCUSSION THREAD

**Discussion by laem** [publicodes/publicodes#562] [General]
Salut ! Je vous expose une problématique qu'on a sur mesaidesréno mais qui vous intéressera je pense. cc @morganmerzouk 

L'un des principes de Mes Aides Réno c'est de permettre aux différents acteurs (privés et publics) d'ajouter le calcul des aides à leurs outils existants, pour diffuser l'information à un maximum de personnes. Ça veut dire par exemple qu'une agence immobilière va intégrer sur son site un petit module de calcul des aides, qui efface une partie de la complexité de mesaidesreno pour apporter une valeur plus directe à ses persona types (ici des acheteurs, pas des propriétaires de long-terme). 

![image](https://github.com/user-attachments/assets/f1f5569b-dcc5-4a2d-bfbb-df1ddb07387f)

Les intégrations possibles ont une diversité importante : à l'opposée du module ci-dessus, on peut imaginer un autre module "mes aides en tant que copropriétaire" qui va aller piocher dans un jeu commun d'aides (MaPrimeRénov' ampleur + aide copro + aide copro individuelle). 

Concrètement, on a un modèle publicodes qui contient des variables "aides=somme des aides", "MaPrimeRénov = une aide en particulier, parmi des dizaines", et des variables de situation d'entrée (par exemple "êtes-vous propriétaire", "nombre de personnes dans le foyer", etc). 

Bref, du gros classique de publicodes quoi :) 

La question qu'on se pose : 
- est-ce qu'on fournit une API sur mesaidesreno.fr/api ? 
- ou plutôt un modèle de calcul versionné et des kits d'intégration (JS via NPM, API à faire tourner à la maison via Express, et potentiellement d'autres langages en fonction des demandes) ? 

Pourquoi la question se pose ? 

Je vais simplifier un peu les choses mais le fond reste réel. Supposons qu'on ait un premier utilisateur A de notre API en septembre 2024. Il développe son application et utilise la variable "aides d'ampleur" qui en septembre vaut aides = somme[MaPrimeRénov' accompagnée, aide à l'audit energétique] et qui renseigne en entrée 3 variables : nombre de personnes du foyer, revenu du foyer, et DPE visé.

Sauf qu'on développe déjà une nouvelle version du modèle plus complète qui va intégrer plusieurs aides supplémentaires, qui viendront avec leur lot de questions en plus : prix d'achat du logement, location longue durée souhaitée, et commune. 

Dans son interface, la somme des aides qui donnait alors 45 000 € pour le persona 1, va donner avec le nouveau modèle 67 000 € et son interface ne sera pas en mesure d'expliquer qu'une nouvelle aide a été intégrée (il n'est pas envisageable de leur demander d'intégrer un parseur publicodes ni la documentation de base actuelle). Ensuite, en janvier on intègre les nouvelles dispositions règlementaires qui pourraient supprimer telle aide (l'aide aux panneaux solaires) et en ajouter une (l'aide aux plaques isolantes pour l'été) --> en fonction de comment le partenaire a codé son intégration, ça peut amocher l'interface mais aussi la casser si les bons `if` et `try catch` ne sont pas présents...

Le potentiel de casse est d'autant plus élevé que les partenaires utilisent des variables internes au moteur. Si l'on veut être tranquille, on n'expose que quelques variables publiques, et on met le reste en privé. Mais ça implique soit que le potentiel d'explication sur les applications de nos partenaires est faible (par exemple, ils pourront pas expliquer que l'aide 3 n'est pas disponible à cause de sa condition 2 qui est invalide pour ce persona), soit qu'on gèle la structure des ces aides et qu'on risque fortement d'accumuler de la dette, alors qu'on est encore qu'en phase de croissance du service. Donc peut-être qu'à terme il sera possible pour nous de limiter les changements à disons une nouvelle version par an, ce qui rend le service versionné de suite plus raisonnable. 

Il nous semble donc impossible de proposer une API sans la versionner : A utilise l'API mesaidesreno.fr/api/1.0.0 et continue à l'utiliser en attendant de prévoir dans sa feuille de route le dev nécessaire pour intégrer les nouveautés et passer à la 1.1.0 ou la 2.0.0 selon si on considère cet ajout d'aides comme un changement sémantique majeur ou pas. 

L'implication pour nous est simple : on doit maintenir plein de versions de l'API en ligne. Techniquement, c'est faisable : il suffit qu'on publie des tags git et qu'on ait sur notre serveur en mémoire tous les modèles compilés, et une API qui instancie le modèle x.y.z en fonction de l'URL demandée par l'utilisateur. Je sais pas trop comment gérer ça en JS/NPM mais il doit bien y avoir une façon de faire, quitte à le faire hors package.json avec un simple script node qui va télécharger l'ensemble des versions d'un paquet. 

Mais est-ce une bonne idée ? On se retrouverait alors à avoir promis la disponibilité de nombreux modèles, potentiellement 1 par semaine, la plupart étant jamais utilisés. Se pose aussi la question du couple mesaidesreno x publicodes. On va intégrer à un moment donné des nouvelles version de publicodes, donc une verison de mesaidesreno est adossée à une version de publicodes. Plus difficile à gérer, ça. On peut bien sûr envisager de déprécier toutes les versions au bout d'un temps x ou d'une différence de version y avec la prod. 

L'alternative, c'est donc de dire : on a une API dernier cri en ligne, mais elle n'est pas versionnée. Dès que vous avez un proto fonctionnel sur cette API qui vous permet d'itérer facilement, vous pouvez vous saisir de notre paquet NPM qui fait modèle + API en Express à mettre sur l'un de vos serveurs, ou à déployer sur un PaaS en 10 minutes. Vous gérez les montées de version. 

En parallèle, on fournit aussi [des iframes](https://mesaidesreno.beta.gouv.fr/module) qui eux, sont toujours à jour, mais ne permettent pas une intégration fine dans la plateforme des partenaires. 

À noter qu'un giga acteur comme Facebook a bien sûr adopté la solution ["on a 20 versions d'API en ligne"](https://developers.facebook.com/docs/graph-api/changelog/version20.0), mais disons qu'ils ont les moyens de le faire :D 

Alors, est-ce que vous avez été confronté à ces questions ? Je pense notamment à @johangirod et mon-entreprise.fr, dont je constate que le paquet NPM n'est plus mis à jour depuis 5 ans, et l'API n'est proposée qu'en v1. La documentation semble dire que [la rétro-compatibilité est donc assurée depuis le début](https://mon-entreprise.urssaf.fr/d%C3%A9veloppeur/api). 

Je ping aussi @EmileRolley bien sûr et @jdauphant qui pourra être intéressé :)

**Comment by laem** [comment]
Bien sûr, une certitude : quelle que soit la solution qu'on propose, versionner notre modèle sémantiquement est un incontournable :)

**Comment by EmileRolley** [comment]
Hello, merci d'ouvrir une discussion sur le sujet.

Je pense que cette question de gestion des versions et d'abstraction des modèles dans dans l'objectif d'être réutilisé dans d'autres projets est importante et pourrait être priorisée pour le prochain publicodes café (le 19/09/2024 à 14h30).

> Je sais pas trop comment gérer ça en JS/NPM mais il doit bien y avoir une façon de faire, quitte à le faire hors package.json avec un simple script node qui va télécharger l'ensemble des versions d'un paquet.

Tu peux regarder ce que j'avais pour l'[API de Nos Gestes Climat](https://github.com/incubateur-ademe/nosgestesclimat-api) même si je pense que stocker les versions en dur n'est pas le plus intelligent (tu peux aussi regarder ce que j'ai fait pour [`catala-dsfr`](https://github.com/catalaLang/catala-dsfr?tab=readme-ov-file#versioned-assets) pour utiliser les alias npm qui me semble être une meilleure approche).

Je pense qu'il y a peut-être également un enjeu sur le fait d'avoir une abstraction au-dessus du modèle (un wrapper JS/endpoint API) pour permettre une meilleure résilience aux changements d'implémentation du modèle et une meilleure DX.

**Comment by johangirod** [comment]
Sur mon-entreprise, on s'est posé exactement les mêmes questions au moment de publier notre API REST et de passer le paquet de règles `modele-social` en v1 : quid des breaking ? On a pris la décision de ne plus en faire. C'était la seule solution qui nous apparaissait viable pour encourager une réutilisation et ne pas avoir à gérer plusieurs versions d'API. 

Depuis, le paquet a été mis à jour très régulièrement **sans aucun breaking**. 

Voici les principes appliqués : 

### 1. Délimiter la surface d'API la plus petite possible

Avec l'attribut `privé`, on peut empêcher certains espaces de nom d'être accessibles depuis l'exterieur. Il y a également une proposition pour avoir un flag de compilation qui mette toutes les règles à privé par défaut, sauf si elles sont rendues explicitement publiques #554

### 2. Ajouter des règles « experimentale »

C'est un flag à ajouter sur les espaces de nom qui permet de les exclure du semver. Ces espaces de noms sont fournis à titre d'experimentation, et peuvent être déplacés / supprimés sans changement de version majeur. Cela prend la forme de warning dans la console / retourné par l'API lorsque l'on demande leur évaluation directement.

### 3. Ne pas supprimer de règles

Lorsque l'on veut changer le nom de règles, on créé des indirections depuis les anciennes. Ainsi si on renomme `aide panneau solaire` en `aide . panneau solaire`, il faudra écrire : 

```
aide panneau solaire: 
  valeur: aide . panneau solaire
  déprécié: utiliser aide . panneau solaire à la place
```

Pour l'instant le moteur n'interprète pas `déprécié` mais on pourrait imaginer l'avoir dans le LSP / en warning à la compilation. On peut aussi creuser le [principe d'alias](#471) proposé par @JalilArfaoui 

### 4. Avoir une interface générée automatiquement (ou des tests de non regression).

Les règles dictées jusqu'à présent permettent des montées de version mineur sans changement cassant. Mais pour que l'interface du client soit à jour, il faut que cette dernière renseigne les bons champs. 

Ainsi, si on ajoute une aide qui dépend d'une nouvelle règle, par exemple « revenu fiscal » il faut que l'interface demande cette valeur à l'utilisateur. 

Le client doit donc prendre en compte que cette liste de question est sujette à évoluer. Il doit donc se baser sur l'attribut `missingVariable`, et aller chercher les informations supplémentaires (question, description, etc.) avec `getRule` (ou la route `rule`). 

Cette partie mérite d'être améliorée (outillage, meilleure prise en compte des types d'entrée et de retour, etc.). La discussion d'Émile va dans ce sens #554

### 0. Profiter du passage à la v1 pour réorganiser et nettoyer le modèle (edit)

On a rarement l'occasion de se poser sur ce qui est déjà écrit, c'est l'occasion de faire un ménage de printemps et de réorganiser la base de code (namespace, règles), pour être uniforme, compréhensible et bien découpée. Cela permettra d'ajouter des fonctionnalités plus facilement par la suite.

---

Gros sujet, mais ça vaut le coup d'en discuter pour dégager des guidelines communes que l'on pourra publier sur la doc / encoder dans un cli.

**Reply by laem** [reply]
aaaah j'avais oublié l'existence de modèle-social, je suis tombé sur https://www.npmjs.com/package/mon-entreprise. 

Je pense qu'il serait bon de marquer ce dernier comme déprécie avec un avertissement bien visible en haut du README @JalilArfaoui

**Reply by laem** [reply]
Merci pour ta réponse très complète ! En effet ça me rassure sur le fait qu'on a peut-être surestimé le nombre de changements cassants qu'on aura dans le futur. 

Ce sont des points très intéressants qu'on va pouvoir intégrer dans notre doc.

**Comment by WoWo79** [comment]
Bonjour à tous !
Je suis dev back sur Agir, et client de différents publicodes à ce jour (NGC, aides velos, aides reno, retrofit)

Je vous pousse ma petite pierre à vos réflexions, et ceci concerne la nature des APIs mentionnées dans cette question :)

- aujourd'hui ces APIs WEB transposent l'API javascript du publicode (getRule, evaluate, missing variables, etc)
- ces APIs ne parlent donc pas le métier de vos domaines sous jacents : par exemple une typologie d'aides, des familles de gestes, un droit, une cotisation, etc
- inversement elles exposent les détails de la modélisation publicode sous jacente.

Mon intuition qui peut éclairer de manière complémentaire le sujet : l'absence d'une abstraction **métier** renforce le risque de ruptures de contrat pour les clients, aggrave aussi votre propre inconfort en tant que développeur de faire évoluer "librement" et aisément votre publicode.

De ma fenêtre modéliser du métier c'est faire émerger des concepts spécifiques au domaine, de nommer ces concepts, de les rendre évidents ("in your face" pour un client et un développeur ^^)
Une modélisation métier fait aussi émerger des invariants structurels ainsi que des relations entres les différents concepts. Ces éléments sont généralement stables et permettent ainsi d'accueillir les changements à venir plus souplement (le métier, s'il est bien capturé, ne change pas du tout au tout le temps passant)

Inversement et par exemple, le fait de rajouter des notions d'"expérimental", des variables "privées" est peut-être un symptôme de vouloir rendre public un objet qui par nature ne s'y prête pas totalement (le publicode et son API javascript/WEB actuelle).

Sinon pour ce qui concerne les APIs en général et leur versionning, il existe en effet de multiples stratégies et pratiques
- du versionning global (tout le domaine métier)
- du versionning granulaire (des ressources partiulières)
- du 'no versionning', avec pour règle d'or de ne JAMAIS casser le contrat ^^
- de la mise à disposition de versions en parallèle (généralement 2-3 max (pas 20 hein))
- la mise à disposition de bacs à sable pour avoir une preview d'une prochaine version majeur et anticiper les impacts
- des releases notes très explicites sur les éléments qui sont mis à jour, voire casssés (le semantique versioning c'est bien, avec un suivi détaillé des impacts c'est mieux)
- des SLA qui indiquent au client combien de temps une version "deprecated" reste disponible en production, ainsi que le rythme global d'évolution anticipé (comme ça le client peut correctement évaluer l'impact de devenir client de cette API)

Enfin, et sur un aspect annexe à la question de départ, j'ai une intuition qu'avoir un API métier peut être une vrai proposition de valeur dans certains cas pour les clients comparé à une API générique orientée DSL publicode, dans le sens d'une amélioration de l'expérience développeur (et expérience client tout court) Et ça je pense que ça peut se tester si on a envie

**Reply by johangirod** [reply]
Salut WoWo79,

Je suis d'accord avec ton argumentaire, avoir une API métier améliore la conceptualisation, et facilite la réutilisation dans une application tierce. En revanche, je nuancerai sur trois points :

-   **Publicodes et modélisation métier** : dans ma vision, publicodes devrait être capable de proposer une API métier « in your face », sans forcément nécessiter de surcouche. Un paquet publicodes bien conçu exposerait uniquement des règles de haut niveau, laissant les détails d'implémentation dans des variables privées (ou non publique). En l'état actuel, je reconnais qu'il manque des abstractions au langage et à l'API pour être utilisé en ce sens (cf #554 et #561)

-   **Facilité d'utilisation** : je suis d'accord qu'une API bien conçue, documentée, réfléchie, sur-mesure vis-à-vis du métier sera toujours mieux qu'une API générique publicodes. En revanche, cela demande également beaucoup plus de travail de conception et de maintien. Par ailleurs, une API publicodes native possède aussi des qualités qui facilitent son utilisation. Elle est découvrable, auto-documentée, uniforme et prévisible.

-  **Abstraction métier pour empêcher les ruptures de contrat** . Cet argument tient dans un contexte où les paquets publicodes ne sont pas publiés, mais seulement leur abstraction métier. Or, la principale raison de l'essaimage de publicodes dans divers projets est la possibilité de réutiliser des règles déjà publiées dans un nouveau modèle, via le système d'import de règles. Les enjeux rencontrés ici sont strictement les mêmes : on souhaite éviter au maximum les ruptures de contrat, ou le cas échéant pouvoir les prévenir. Il faudra donc **dans tous les cas** adresser ces problématiques au niveau des règles publicodes, sauf à voir disparaître l'interopérabilité des modèles dans l'écosystème.


Mon point de vue est donc le suivant. Je pense que la réutilisation de modèles publicodes par Agir est une belle opportunité de faire évoluer le langage afin que le développement et la réutilisation des modèles publiés se rapproche au maximum de celui d'une API métier « sur mesure ». 

Ce, dans une optique de simplification, et de rationalisation des coûts (deux APIs à maintenir, cela est plus compliqué qu'une seule).

Cela étant dit, je pense comme toi que ces hypothèses doivent être testées. Je suis donc très favorable à prendre un cas d'usage réel (par exemple, mesaidesréno) et de se plier à l'exercice de concevoir l'API métier idéale (sans forcément l'implémenter).

Cela permettrait de voir s'il est envisageable d'obtenir quelque chose de similaire en publicodes pur et, si oui, d'orienter les futurs développements pour y parvenir.

**Reply by laem** [reply]
Merci pour le retour @WoWo79 ! 

Je pense que la mise à disposition d'une API métier est un sujet relativement différent de ce sujet initial. Mais c'est très bien d'en parler ici aussi, car c'est pas tout à fait étranger. 

Il faut qu'on rentre dans le concret de ce que tu appelles une API métier, car l'API fournie par publicodes (sous forme de la mise à disposition du point d'entrée setSituation(situation).evalute(ruleName)) est une API métier : elle expose de façon directe toutes les variables du modèle. Par contre, c'est une API qui permet de tout faire. Donc si je comprends bien ce que tu propose, ce serait de fournir une API plus restreinte, et c'est là que le lien avec ma question initiale se fait : moins une API est puissante, moins elle a de changements cassants. 

Donc il est important de discuter concrètement de ce que serait cette API restreinte. L'idée fondamentale de publicodes c'est de ne pas créer de couches intermédiaires de code non documenté et en dehors du modèle principal, ou du moins de le réduire au maximum. Merci pour ta réponse Johan, bien d'accord. 

Aussi, l'API ne sera jamais seulement une API Web : la plupart des intégrations se feront avec import direct du modèle pour éviter une dépendance inutile à un serveur tiers et toutes les implications de RGPD, de stabilité et d'UX (latence d'un appel HTTP en plus) qui vont avec. Donc le sujet serait plutôt celui d'une surcouche de fonctions au-dessus de publicodes. 

Reste donc à déterminer quelles sont les fonctions qui seraient utiles à un intégrateur et qui ne sont pas déjà (ou pas correctement, pas facilement) exposées par publicodes directement. 

Par exemple, on pourrait exposer une fonction listQuestions qui donnerait la liste des questions du simulateur principal de MAR (Mes Aides Réno), qui fixerait la règle de calcul à "aides" (celle du simulateur principal), mais ce ne serait qu'un helper sur le missingVariables renvoyé directement par publicodes. Il faudrait qu'on spécifie la signature des `question` renvoyées par la fonction, mais ça risquerait d'être moins riche et moins utile que d'aller chercher la définition publicodes de la question via le YAML directement, ou `engine.getParsedRules()[ruleName]`. 

@WoWo79 est-ce que vous avez déjà identifié des fonctions d'aide qu'on pourrait intégrer dans une surcouche d'API car spécifique à MAR, ou à l'inverse non spécifiques et qui viendraient corriger des manques de publicodes et qu'on pourrait du coup intégrer directement dans publicodes pour améliorer l'xp dev pour tous ? 

On avait parlé de la capacité à deviner les seuils des questions pour le revenu (aussi utilisé sur Mes Aides Vélo), mais c'est pas trivial à généraliser (cf notre [composant](https://github.com/betagouv/reno/blob/master/components/SmartInput.tsx)), et on avait conclu que vous n'en auriez pas besoin dans un premier temps.

**Reply by laem** [reply]
Pour avancer sur le sujet en parallèle, @WoWo79 je pense que dans l'équipe MAR on doit aussi proposer des démos de réutilisation de notre paquet MAR. Donc publier un paquet NPM qui expose le JSON, et en faire une démo (en Next pour commencer) sur un cas d'usage précis. Comme ça on sera confronté aux mêmes problématiques qu'une réutilisateur qui ne se plongerait pas dans le code de MAR.  À suivre ici https://github.com/betagouv/reno/issues/218

**Reply by laem** [reply]
MAJ : voilà le lien d'un codesandbox ultra basique, qu'on va améliorer dès que possible pour notamment proposer un formulaire. https://codesandbox.io/p/devbox/reno-demo-xzs68w

**Reply by EmileRolley** [reply]
>  est-ce que vous avez déjà identifié des fonctions d'aide qu'on pourrait intégrer dans une surcouche d'API car spécifique à MAR, ou à l'inverse non spécifiques et qui viendraient corriger des manques de publicodes et qu'on pourrait du coup intégrer directement dans publicodes pour améliorer l'xp dev pour tous

Intéressé par cette question également ! :eyes: 

> Donc publier un paquet NPM qui expose le JSON, et en faire une démo.

Si vous vous penchez sur le sujet de la publication des règles Publicodes, je pense qu'il ce serait le moment de réfléchir à un build system ou a minima à un format de paquet pour qu'il puisse facilement être utiliser par le système d'import (qui attend actuellement un `<package-name>.model.json` à la racine du paquet). 

Je ne sais pas si c'est voulu mais le paquet `mesaidesreno` sur NPM contient tout le code source du repo, il faut préciser un champs `files` dans le `package.json` pour inclure uniquement le fichier JSON. N'hésites pas à regarder le template https://github.com/publicodes/model-template il est fait pour ça (même si comme expliqué plus haut, je pense qu'il serait temps de le remplacer par un build system avec une CLI et un script create pour npm).

**Reply by laem** [reply]
Ah oui merci Émile, je suis allé vite hier pour avoir un premier paquet utilisable (et utilisé :heavy_check_mark:) mais tu as raison, il faut faire ça proprement :)

**Comment by WoWo79** [comment]
hahaha, moi qui adore les débats technico-sémantique, je vais devoir me restreindre ^^
Ou alors on peut aussi lancer un autre thread en // sur la nature profonde de publicodes et de son API javascript \o/

Sinon en effet pour être concret : 

- je bosse justement en ce moment sur l'intégration, en tant que client, du publicode aide-reno
- à l'occasion de cette intégration je me propose de faire l'effort d'émerger quelques abstractions de plus haut niveau par dessus l'engine, celles qui me paraissent faire sens
- je construis également une API WEB REST orienté métier pour le mobile et le web pour manipuler/simuler les aides reno (ça c'est notre philosophie d'architecture Agir, ce n'est pas un effort particulier)
- je vous reverse en connaissance/feedback ces deux éléments (API WEB + abstractions JS), voire peut-être des éléments de mon domaine métier interne, si pertinent
- cerise sur le gateau, ayant tout cela en main, je serais curieux de faire ensuite des tests utilisateurs (ici utilisateurs = développeurs client des aides réno)
  - trouver quelques développeurs volontaires pour se prêter au jeu
  - d'une part leur proposer une API fonctionnelle (telle que je vais produire)
  - d'autre par leur proposer le publicode et son API javascript (ou WEB)
  - demander de réaliser quelques usescases basiques en scénarios de tests
  - observer leur capacité à faire, les frictions, recueillir les feedbacks

<img width="489" alt="archi-integration-aide-reno" src="https://github.com/user-attachments/assets/c3a61307-22b4-4488-924f-9c5f8b64ab30">
